### PR TITLE
Remove L-W ixn group and merge w/P-L group

### PR DESCRIPTION
**Pull request by jkausrelay**
- Remove the L-W specific ixn group and merge it back in with the P-L (now env-L) ixn gorup.
- This should improve performance
- Maintain backward compatibility with FFs that have a separate 'solvent' section if the parameters match those of the original ixn group

### CODE REVIEW

**Comment by badisa** [comment]
Could you post the benchmarks for timemachine before and after this change?

File: timemachine/fe/single_topology.py [hunk 1]
```diff
@@ -1869,8 +1869,7 @@ def _parameterize_host_guest_nonbonded_ixn(
         num_guest_atoms = self.get_num_atoms()
         cutoff = host_nonbonded.potential.cutoff
 
-        guest_ixn_water_params = self._get_guest_params(self.ff.q_handle_solv, self.ff.lj_handle_solv, lamb, cutoff)
-        guest_ixn_other_params = self._get_guest_params(self.ff.q_handle, self.ff.lj_handle, lamb, cutoff)
+        guest_ixn_env_params = self._get_guest_params(self.ff.q_handle, self.ff.lj_handle, lamb, cutoff)
```
Comment (badisa) [inline timemachine/fe/single_topology.py:1872]: Should this be `q_handle_intra`?

File: timemachine/ff/__init__.py [hunk 2]
```diff
@@ -230,17 +220,14 @@ def from_handlers(
...
-            elif isinstance(q_handle, nonbonded.AM1BCCHandler):
-                q_handle_solv = nonbonded.AM1BCCSolventHandler(q_handle.smirks, q_handle.params, q_handle.props)
-            elif isinstance(q_handle, nonbonded.SimpleChargeHandler):
-                q_handle_solv = nonbonded.SimpleChargeSolventHandler(q_handle.smirks, q_handle.params, q_handle.props)
-            else:
-                raise ValueError(f"Unsupported charge handler {q_handle}")
+        if q_handle_solv is not None:
+            assert q_handle is not None
+            assert q_handle_solv is not None
+            assert_almost_equal(
```
Comment (badisa) [inline timemachine/ff/__init__.py:226]: Should this be `assert_equal`?

**Review by badisa — APPROVED** [review]

File: tests/common.py [hunk 3]
```diff
@@ -533,31 +519,6 @@ def check_split_ixns(
...
-        )
-
-        # U_solv = U_sum_ref - WL_ref + WL_solv
-        expected_u = sum_u_ref - WL_u_ref + WL_u_solv
-        expected_grad = sum_grad_ref - WL_grad_ref + WL_grad_solv
-
-        np.testing.assert_allclose(expected_u, sum_u_solv, rtol=rtol, atol=atol)
-        np.testing.assert_allclose(expected_grad, sum_grad_solv, rtol=rtol, atol=atol)
-
         # Compute the grads, potential with the protein-ligand terms scaled
```
Comment (maxentile) [inline tests/common.py:522]: non-obvious loss of test functionality -- should this test be retained, but just replace `ffs.solv` with `ffs.prot` ?

(to assert that `ffs.prot` now controls both the WL and PL interaction group?)

**Comment by jkausrelay** [comment]
> Could you post the benchmarks for timemachine before and after this change?

Before
```
tests/test_benchmark.py::test_dhfr steps per second: 1915.811
ns per day: 413.815
steps per second: 1915.767
ns per day: 413.806
dhfr-apo: N=23558 speed: 413.81ns/day dt: 2.5fs (ran 200 steps in 0.10s)
steps per second: 1806.745
ns per day: 390.257
steps per second: 1811.560
ns per day: 391.297
dhfr-apo-barostat-interval-25: N=23558 speed: 391.30ns/day dt: 2.5fs (ran 200 steps in 0.11s)
PASSED
tests/test_benchmark.py::test_hif2a building a protein system with 1758 protein atoms and 7047 water atoms
Optimizing initial state at λ=0.1
----------------------------------------------------------------------
performing BFGS minimization on 265 atoms
(holding the other 8575 atoms frozen)
U(x_0) = -102393.345
Warning: Desired error not necessarily achieved due to precision loss.
         Current function value: -103237.901388
         Iterations: 364
         Function evaluations: 449
         Gradient evaluations: 435
U(x_final) = -103237.901
atom with highest force norm after minimization: 1117
force(x_final)[1117] = [-223.80926003  121.6185849   150.32131463]
----------------------------------------------------------------------
steps per second: 3509.614
ns per day: 758.077
steps per second: 3501.644
ns per day: 756.355
hif2a-apo: N=8805 speed: 756.36ns/day dt: 2.5fs (ran 200 steps in 0.06s)
steps per second: 3376.105
ns per day: 729.239
steps per second: 3348.639
ns per day: 723.306
hif2a-apo-barostat-interval-25: N=8805 speed: 723.31ns/day dt: 2.5fs (ran 200 steps in 0.06s)
steps per second: 2967.255
ns per day: 640.927
steps per second: 2951.937
ns per day: 637.618
hif2a-rbfe-barostat-interval-25: N=8840 speed: 637.62ns/day dt: 2.5fs (ran 200 steps in 0.07s)
steps per second: 7011.776
ns per day: 1514.544
steps per second: 7065.995
ns per day: 1526.255
hif2a-rbfe-local: N=8840 speed: 1526.25ns/day dt: 2.5fs (ran 200 steps in 0.03s)
Warning::Not running water sampling every batch, interval is too large
steps per second: 2956.108
ns per day: 638.519
steps per second: 2930.293
ns per day: 632.943
hif2a-rbfe-barostat-interval-25-water-sampling-interval-400: N=8840 speed: 632.94ns/day dt: 2.5fs (ran 200 steps in 0.07s)
PASSED
tests/test_benchmark.py::test_solvent Optimizing initial state at λ=0.1
----------------------------------------------------------------------
performing BFGS minimization on 199 atoms
(holding the other 6118 atoms frozen)
U(x_0) = -83324.834
Warning: Desired error not necessarily achieved due to precision loss.
         Current function value: -84121.528281
         Iterations: 484
         Function evaluations: 587
         Gradient evaluations: 572
U(x_final) = -84121.528
atom with highest force norm after minimization: 645
force(x_final)[645] = [ 313.70727607  -35.58211722 -333.58607851]
----------------------------------------------------------------------
steps per second: 5631.374
ns per day: 1216.377
steps per second: 5600.807
ns per day: 1209.774
solvent-apo: N=6282 speed: 1209.77ns/day dt: 2.5fs (ran 200 steps in 0.04s)
steps per second: 5095.988
ns per day: 1100.733
steps per second: 5073.428
ns per day: 1095.860
solvent-apo-barostat-interval-25: N=6282 speed: 1095.86ns/day dt: 2.5fs (ran 200 steps in 0.04s)
steps per second: 4191.370
ns per day: 905.336
steps per second: 4187.667
ns per day: 904.536
solvent-rbfe-barostat-interval-25: N=6317 speed: 904.54ns/day dt: 2.5fs (ran 200 steps in 0.05s)
steps per second: 6264.830
ns per day: 1353.203
steps per second: 6338.248
ns per day: 1369.062
solvent-rbfe-local: N=6317 speed: 1369.06ns/day dt: 2.5fs (ran 200 steps in 0.03s)
PASSED
tests/test_benchmark.py::test_vacuum Optimizing initial state at λ=0.1
----------------------------------------------------------------------
performing BFGS minimization on 35 atoms
(holding the other 0 atoms frozen)
U(x_0) = 891.203
Warning: Desired error not necessarily achieved due to precision loss.
         Current function value: 664.658141
         Iterations: 150
         Function evaluations: 274
         Gradient evaluations: 261
U(x_final) = 664.658
atom with highest force norm after minimization: 18
force(x_final)[18] = [-0.73993241 -0.04782357  0.28527757]
----------------------------------------------------------------------
steps per second: 38370.725
ns per day: 8288.077
steps per second: 38836.148
ns per day: 8388.608
vacuum-rbfe: N=35 speed: 8388.61ns/day dt: 2.5fs (ran 200 steps in 0.01s)
PASSED
tests/test_benchmark.py::test_nonbonded_interaction_group_potential building a protein system with 1758 protein atoms and 7047 water atoms
Optimizing initial state at λ=0.1
----------------------------------------------------------------------
performing BFGS minimization on 265 atoms
(holding the other 8575 atoms frozen)
U(x_0) = -102393.345
Warning: Desired error not necessarily achieved due to precision loss.
         Current function value: -103237.901388
         Iterations: 364
         Function evaluations: 449
         Gradient evaluations: 435
U(x_final) = -103237.901
atom with highest force norm after minimization: 1117
force(x_final)[1117] = [-223.80926003  121.6185849   150.32131463]
----------------------------------------------------------------------
NonbondedInteractionGroup_f32: N=8840 Frames=100 Params=5 speed: 1033.45 executions/seconds (ran 1000 potentials in 0.97s) du_dp=True, du_dx=True, u=True
NonbondedInteractionGroup_f64: N=8840 Frames=100 Params=5 speed: 514.39 executions/seconds (ran 1000 potentials in 1.94s) du_dp=True, du_dx=True, u=True
PASSED
tests/test_benchmark.py::test_hif2a_potentials HarmonicBond_f32: N=8840 Frames=100 Params=5 speed: 1360.92 executions/seconds (ran 1000 potentials in 0.73s) du_dp=True, du_dx=True, u=True
HarmonicBond_f64: N=8840 Frames=100 Params=5 speed: 1357.16 executions/seconds (ran 1000 potentials in 0.74s) du_dp=True, du_dx=True, u=True
HarmonicAngleStable_f32: N=8840 Frames=100 Params=5 speed: 1313.81 executions/seconds (ran 1000 potentials in 0.76s) du_dp=True, du_dx=True, u=True
HarmonicAngleStable_f64: N=8840 Frames=100 Params=5 speed: 1293.55 executions/seconds (ran 1000 potentials in 0.77s) du_dp=True, du_dx=True, u=True
PeriodicTorsion_f32: N=8840 Frames=100 Params=5 speed: 1317.89 executions/seconds (ran 1000 potentials in 0.76s) du_dp=True, du_dx=True, u=True
PeriodicTorsion_f64: N=8840 Frames=100 Params=5 speed: 1296.50 executions/seconds (ran 1000 potentials in 0.77s) du_dp=True, du_dx=True, u=True
ChiralAtomRestraint_f32: N=8840 Frames=100 Params=5 speed: 1538.59 executions/seconds (ran 1000 potentials in 0.65s) du_dp=True, du_dx=True, u=True
ChiralAtomRestraint_f64: N=8840 Frames=100 Params=5 speed: 1516.91 executions/seconds (ran 1000 potentials in 0.66s) du_dp=True, du_dx=True, u=True
NonbondedPairListPrecomputed_f32: N=8840 Frames=100 Params=5 speed: 1520.78 executions/seconds (ran 1000 potentials in 0.66s) du_dp=True, du_dx=True, u=True
NonbondedPairListPrecomputed_f64: N=8840 Frames=100 Params=5 speed: 1492.61 executions/seconds (ran 1000 potentials in 0.67s) du_dp=True, du_dx=True, u=True
Nonbonded_f32: N=8840 Frames=100 Params=5 speed: 415.38 executions/seconds (ran 1000 potentials in 2.41s) du_dp=True, du_dx=True, u=True
Nonbonded_f64: N=8840 Frames=100 Params=5 speed: 21.11 executions/seconds (ran 1000 potentials in 47.37s) du_dp=True, du_dx=True, u=True
SummedPotential(NonbondedInteractionGroup)_f32: N=8840 Frames=100 Params=5 speed: 1021.51 executions/seconds (ran 1000 potentials in 0.98s) du_dp=True, du_dx=True, u=True
SummedPotential(NonbondedInteractionGroup)_f64: N=8840 Frames=100 Params=5 speed: 475.71 executions/seconds (ran 1000 potentials in 2.10s) du_dp=True, du_dx=True, u=True
PASSED

======================== 6 passed in 132.21s (0:02:12) =========================
```

After
```
tests/test_benchmark.py::test_dhfr steps per second: 1913.538
ns per day: 413.324
steps per second: 1916.139
ns per day: 413.886
dhfr-apo: N=23558 speed: 413.89ns/day dt: 2.5fs (ran 200 steps in 0.10s)
steps per second: 1735.478
ns per day: 374.863
steps per second: 1752.929
ns per day: 378.633
dhfr-apo-barostat-interval-25: N=23558 speed: 378.63ns/day dt: 2.5fs (ran 200 steps in 0.11s)
PASSED
tests/test_benchmark.py::test_hif2a building a protein system with 1758 protein atoms and 7047 water atoms
Optimizing initial state at λ=0.1
----------------------------------------------------------------------
performing BFGS minimization on 265 atoms
(holding the other 8575 atoms frozen)
U(x_0) = -102393.345
Warning: Desired error not necessarily achieved due to precision loss.
         Current function value: -103237.901388
         Iterations: 364
         Function evaluations: 449
         Gradient evaluations: 435
U(x_final) = -103237.901
atom with highest force norm after minimization: 1117
force(x_final)[1117] = [-223.80926003  121.6185849   150.32131463]
----------------------------------------------------------------------
steps per second: 3517.325
ns per day: 759.742
steps per second: 3508.322
ns per day: 757.798
hif2a-apo: N=8805 speed: 757.80ns/day dt: 2.5fs (ran 200 steps in 0.06s)
steps per second: 3169.843
ns per day: 684.686
steps per second: 3164.366
ns per day: 683.503
hif2a-apo-barostat-interval-25: N=8805 speed: 683.50ns/day dt: 2.5fs (ran 200 steps in 0.06s)
steps per second: 2972.238
ns per day: 642.004
steps per second: 2987.641
ns per day: 645.330
hif2a-rbfe-barostat-interval-25: N=8840 speed: 645.33ns/day dt: 2.5fs (ran 200 steps in 0.07s)
steps per second: 6939.961
ns per day: 1499.031
steps per second: 7028.873
ns per day: 1518.236
hif2a-rbfe-local: N=8840 speed: 1518.24ns/day dt: 2.5fs (ran 200 steps in 0.03s)
Warning::Not running water sampling every batch, interval is too large
steps per second: 2980.412
ns per day: 643.769
steps per second: 2932.854
ns per day: 633.496
hif2a-rbfe-barostat-interval-25-water-sampling-interval-400: N=8840 speed: 633.50ns/day dt: 2.5fs (ran 200 steps in 0.07s)
PASSED
tests/test_benchmark.py::test_solvent Optimizing initial state at λ=0.1
----------------------------------------------------------------------
performing BFGS minimization on 199 atoms
(holding the other 6118 atoms frozen)
U(x_0) = -83324.834
Warning: Desired error not necessarily achieved due to precision loss.
         Current function value: -84121.528281
         Iterations: 484
         Function evaluations: 587
         Gradient evaluations: 572
U(x_final) = -84121.528
atom with highest force norm after minimization: 645
force(x_final)[645] = [ 313.70727607  -35.58211722 -333.58607851]
----------------------------------------------------------------------
steps per second: 5673.499
ns per day: 1225.476
steps per second: 5654.301
ns per day: 1221.329
solvent-apo: N=6282 speed: 1221.33ns/day dt: 2.5fs (ran 200 steps in 0.04s)
steps per second: 5121.438
ns per day: 1106.231
steps per second: 5093.884
ns per day: 1100.279
solvent-apo-barostat-interval-25: N=6282 speed: 1100.28ns/day dt: 2.5fs (ran 200 steps in 0.04s)
steps per second: 4221.448
ns per day: 911.833
steps per second: 4202.899
ns per day: 907.826
solvent-rbfe-barostat-interval-25: N=6317 speed: 907.83ns/day dt: 2.5fs (ran 200 steps in 0.05s)
steps per second: 6317.294
ns per day: 1364.535
steps per second: 6351.782
ns per day: 1371.985
solvent-rbfe-local: N=6317 speed: 1371.98ns/day dt: 2.5fs (ran 200 steps in 0.03s)
PASSED
tests/test_benchmark.py::test_vacuum Optimizing initial state at λ=0.1
----------------------------------------------------------------------
performing BFGS minimization on 35 atoms
(holding the other 0 atoms frozen)
U(x_0) = 891.203
Warning: Desired error not necessarily achieved due to precision loss.
         Current function value: 664.658141
         Iterations: 150
         Function evaluations: 274
         Gradient evaluations: 261
U(x_final) = 664.658
atom with highest force norm after minimization: 18
force(x_final)[18] = [-0.73993241 -0.04782357  0.28527757]
----------------------------------------------------------------------
steps per second: 40329.846
ns per day: 8711.247
steps per second: 40587.420
ns per day: 8766.883
vacuum-rbfe: N=35 speed: 8766.88ns/day dt: 2.5fs (ran 200 steps in 0.01s)
PASSED
tests/test_benchmark.py::test_nonbonded_interaction_group_potential building a protein system with 1758 protein atoms and 7047 water atoms
Optimizing initial state at λ=0.1
----------------------------------------------------------------------
performing BFGS minimization on 265 atoms
(holding the other 8575 atoms frozen)
U(x_0) = -102393.345
Warning: Desired error not necessarily achieved due to precision loss.
         Current function value: -103237.901388
         Iterations: 364
         Function evaluations: 449
         Gradient evaluations: 435
U(x_final) = -103237.901
atom with highest force norm after minimization: 1117
force(x_final)[1117] = [-223.80926003  121.6185849   150.32131463]
----------------------------------------------------------------------
NonbondedInteractionGroup_f32: N=8840 Frames=100 Params=5 speed: 1034.30 executions/seconds (ran 1000 potentials in 0.97s) du_dp=True, du_dx=True, u=True
NonbondedInteractionGroup_f64: N=8840 Frames=100 Params=5 speed: 514.39 executions/seconds (ran 1000 potentials in 1.94s) du_dp=True, du_dx=True, u=True
PASSED
tests/test_benchmark.py::test_hif2a_potentials HarmonicBond_f32: N=8840 Frames=100 Params=5 speed: 1359.53 executions/seconds (ran 1000 potentials in 0.74s) du_dp=True, du_dx=True, u=True
HarmonicBond_f64: N=8840 Frames=100 Params=5 speed: 1357.91 executions/seconds (ran 1000 potentials in 0.74s) du_dp=True, du_dx=True, u=True
HarmonicAngleStable_f32: N=8840 Frames=100 Params=5 speed: 1312.71 executions/seconds (ran 1000 potentials in 0.76s) du_dp=True, du_dx=True, u=True
HarmonicAngleStable_f64: N=8840 Frames=100 Params=5 speed: 1294.29 executions/seconds (ran 1000 potentials in 0.77s) du_dp=True, du_dx=True, u=True
PeriodicTorsion_f32: N=8840 Frames=100 Params=5 speed: 1319.19 executions/seconds (ran 1000 potentials in 0.76s) du_dp=True, du_dx=True, u=True
PeriodicTorsion_f64: N=8840 Frames=100 Params=5 speed: 1296.65 executions/seconds (ran 1000 potentials in 0.77s) du_dp=True, du_dx=True, u=True
ChiralAtomRestraint_f32: N=8840 Frames=100 Params=5 speed: 1539.39 executions/seconds (ran 1000 potentials in 0.65s) du_dp=True, du_dx=True, u=True
ChiralAtomRestraint_f64: N=8840 Frames=100 Params=5 speed: 1516.67 executions/seconds (ran 1000 potentials in 0.66s) du_dp=True, du_dx=True, u=True
NonbondedPairListPrecomputed_f32: N=8840 Frames=100 Params=5 speed: 1520.35 executions/seconds (ran 1000 potentials in 0.66s) du_dp=True, du_dx=True, u=True
NonbondedPairListPrecomputed_f64: N=8840 Frames=100 Params=5 speed: 1491.70 executions/seconds (ran 1000 potentials in 0.67s) du_dp=True, du_dx=True, u=True
Nonbonded_f32: N=8840 Frames=100 Params=5 speed: 413.76 executions/seconds (ran 1000 potentials in 2.42s) du_dp=True, du_dx=True, u=True
Nonbonded_f64: N=8840 Frames=100 Params=5 speed: 21.14 executions/seconds (ran 1000 potentials in 47.30s) du_dp=True, du_dx=True, u=True
SummedPotential(NonbondedInteractionGroup)_f32: N=8840 Frames=100 Params=5 speed: 1019.28 executions/seconds (ran 1000 potentials in 0.98s) du_dp=True, du_dx=True, u=True
SummedPotential(NonbondedInteractionGroup)_f64: N=8840 Frames=100 Params=5 speed: 474.97 executions/seconds (ran 1000 potentials in 2.11s) du_dp=True, du_dx=True, u=True
PASSED
```

File: tests/test_single_topology.py [hunk 4]
```diff
@@ -854,7 +854,7 @@ def get_vacuum_solvent_u_grads(ff, lamb):
             vacuum_u_inter_scaled,
             solvent_grad_inter_scaled,
             solvent_u_inter_scaled,
-        ) = get_vacuum_solvent_u_grads(ffs.solv, lamb)
+        ) = get_vacuum_solvent_u_grads(ffs.prot, lamb)
```
Comment (maxentile) [inline tests/test_single_topology.py:857]: correct usage, but could look like a bug when reading over it without fresh context. could consider renaming `ffs.prot` to `ffs.env` or similar

File: tests/common.py [hunk 5]
```diff
@@ -378,7 +378,6 @@ def params_with_w_coords(w_coords):
 class SplitForcefield:
     ref: Forcefield  # ref ff
     intra: Forcefield  # intermolecular charge terms/LJ terms scaled
-    solv: Forcefield  # water-ligand charge terms/LJ terms scaled
     prot: Forcefield  # protein-ligand charge terms/LJ terms scaled
```
Comment (maxentile) [inline tests/common.py:381]: behavior change invalidates name and line comment

**Review by maxentile — APPROVED** [review]

File: tests/test_single_topology.py (hunk 4 above)
Comment (proteneer) [reply tests/test_single_topology.py:857]: +1

File: tests/common.py (hunk 3 above)
Comment (jkausrelay) [reply tests/common.py:522]: Updated the test to make it more clear what's happening here. Now that we only have L-E scaled together, the first part of the test (deleted) is doing the same check as the second, so it's not needed.

**Pull request merged**